neurom/analysis/dendrogram.py

```diff
@@ -0,0 +1,295 @@
+# Copyright (c) 2015, Ecole Polytechnique Federale de Lausanne, Blue Brain Project
+# All rights reserved.
+#
+# This file is part of NeuroM <https://github.com/BlueBrain/NeuroM>
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+#     1. Redistributions of source code must retain the above copyright
+#        notice, this list of conditions and the following disclaimer.
+#     2. Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in the
+#        documentation and/or other materials provided with the distribution.
+#     3. Neither the name of the copyright holder nor the names of
+#        its contributors may be used to endorse or promote products
+#        derived from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY
+# DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+# (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+# ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+'''Dendrogram helper functions and class'''
+
+from neurom.core.tree import Tree, ipreorder
+from neurom.core.neuron import Neuron
+from neurom.analysis.morphtree import n_segments, n_bifurcations, n_terminations
+from neurom.analysis.morphmath import segment_length
+from neurom.core.dataformat import COLS
+
+import numpy as np
+import sys
+
+
+def _recursion_handler(func):
+    ''' Decorator that takes a recursive function which operates on trees or neurons
+    and estimates the max number of calls to the recursive function. This is achieved
+    by finding the max number of nodes of the given tree or from the biggest neurite in
+    the given neuron.
+    '''
+
+    def depth(tree):
+        ''' Returns the depth which is defined as the number of nodes in a tree
+        '''
+        return sum(1 for _ in ipreorder(tree))
+
+    def set_limit(self):
+        ''' Set the recursion limit, run the function and restore its initial value
+        '''
+        neurites = self.obj.neurites if hasattr(self.obj, 'neurites') else [self.obj]
```

Comment:
And this dependency on `self.obj` makes quite a strong coupling between the decorator and the thing being detorated.